[Wrongian] iP

Wrongian opened the pull request:
# Demacia
> "Sometimes the best waifu is the best boy." ~Somebody on the internet

Demacia is a chatbot based on the best boy from the web novel "Game of the World Tree"(NOT SPONSORED BTW)

One of my **favourite** novels of all time, it genuinely has the **best** depiction of a fantasy VRMMO I have seen so far.

Let me actually tell you why its so incredible
1.  The VRMMO is quite realistic
* I dont like skills like "martial arts" and activating skills to execute specific movements because these kinds of skills break some of the fantasy of VRMMOs because they fundamentally restrict what you do in the games. 
* Another reason is that putting more skill points into a skill and suddenly learning how to do a specific type of martial arts is kinda boring? I rather people develop their own martial arts and skills using the game mechanics of the new world like what happens in Game of the World Tree.
* The fundamental difference is that in Game of the World Tree, the players have to adapt to the systems of the game via novel and creative ways to thrive.

2. World/Civilisation development
Over its runtime, you get to see civilisation develop from stick and stones with a couple hundred players into interdimensional travel with hundreds of millions of players. You get to see the perspectives of new players that join at different time periods. You get to see raids of millions of players and the dynamics that go into them.

3. Funny Antics
A new mechanic is added and players try to take advantage of these new developments to the best of their ability through bug abuse, creative thinking, etc. God's punishment bomb abuse and the "priest" class was hilarious.

uhhhh oh yeah the chatbot haha, heres the [link](https://github.com/Wrongian/ip) to the repo

I need to do a ~~limbillion~~ some things:
- [ ] Sleep
- [ ] Do 20 hours of CS2103T per week
- [x] Drink a sane amount of green tea

Heres some code I guess? Why am I randomly providing code? I dunno :wink::wink::wink::wink:

```java
+[----->+++<]>+.++++++++++++..----.+++.+[-->+<]>.-----------..++[->++<]>.+++++++.+++++++++++.--------.-[--->+<]>.-----[++>---<]>.--[--->+<]>-.++++++++++++.--.+++[->+++++<]>-.++++++++++.-.-[->++++++<]>+.+++[->++++<]>.[--->+<]>---.--[----->+<]>+.
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Wrongian committed:
Renamed main class

Wrongian committed:
Added greeting

Wrongian committed:
Changed bot and terminal interactions

Wrongian committed:
Changed the exit function to explicitly exit

Wrongian committed:
Added echo feature

Wrongian committed:
Adding the adding and listing task feature

Wrongian committed:
Adding taskList

Wrongian committed:
Added the Task class

Wrongian committed:
Added the ability to mark and unmark tasks as done

Wrongian committed:
Added utils class for reusable methods

Wrongian committed:
Added support for multiple tasks types

Wrongian committed:
Added tracking of todos, events and deadlines

Wrongian committed:
Fixed spacing issues

Wrongian committed:
Changed name of certain test files

Wrongian committed:
Added automated ui testing

Wrongian committed:
Revert "Added automated ui testing"

This reverts commit 97de758275c2372489721006e832ddcf403f8d12.

Wrongian committed:
Reapply "Added automated ui testing"

This reverts commit 904d0aa29ebc5ff62aa0cc09deed962baee0d898.

Wrongian committed:
Redid command parsing logic

Wrongian committed:
Added error checking for the task array

Wrongian committed:
Added error handling with custom exceptiosn

Wrongian committed:
Added usage requirements when a invalid command is inputted

Wrongian committed:
Changed the data structure to store tasks to be an ArrayList

Wrongian committed:
Added the ability for the user to delete tasks

Wrongian committed:
Added checks to ensure integer arguments are integers

Wrongian committed:
Add additional file error types

When a file being read or written to. There are no exceptions in
the case where a file is not writable or readable.

Not being writable or readable should be considered a checked error

Add not being writable and readable as custom duke exceptions.

Making custom exceptions allow for custom behavior for these errors
that align with the requirements of the project

Wrongian committed:
Add file handler class

Reading and writing to files have to be done separately in each class
that uses it.

The reading and writing to files can be abstracted to avoid code
duplication.

Create a file handler class to abstract these details. Modify he
exceptions handling the read and write permissions to be more helpful.

Less boilerplate needs to be written for file reading and writing.

Wrongian committed:
Revert "Add file handler class"

This reverts commit a0e400c767e2f4ef55d040a9354314270760b6ff.

Wrongian committed:
Reapply "Add file handler class"

This reverts commit bfd8d9db02d6ec3b8f30f064fcc6574065968fb5.

Wrongian committed:
Add the ability to save and load

Users had to enter all the task details and this was not persistent
across user sessions.

Add the ability to save and load. Save files have the format of tasks
being separated by newlines and different attributes of tasks are
separated by commas. The program will parse and unparse this when
saving and loading. Saving and loading is done via the SaveHandler
class. Classes that can be saved implements the Saveable interface. Add
the SaveData class to contain the data that needs to be saved.
Update the text-ui-test tests. Update .gitignore to account for the
new save files. Add InavlidSaveException class to handle exceptions for
invalid formatting of the save file.

The SaveHandler class is to abstract away saving. The SaveData class
is to encapsulate the test data to make updating it in the foture
easier.

Wrongian committed:
Change description of the format to save

Used an older version of the save format as a comment.

Needs to be updated to the newer version of the save format.

Update the comment to reflect the newer save format.

To make it consistent for other people looking at the code and for
future me.

Wrongian committed:
Merge branch 'Branch-Level-7'

Wrongian committed:
Change exceptions to be more user friendly

Exceptions were less descriptive.

Users may not know how to deal with the error.

Change error messages to be more descriptive on what to do.

Any help for the user would be nice, especially in a command line
application.

Wrongian committed:
Merge branch 'Branch-Level-7'

Wrongian committed:
Merge branch 'master' into branch-Level-8

Wrongian committed:
Merge branch 'master' into branch-Level-7

Wrongian committed:
Delete unused functions in Utils class

The method joinStringArray() is unused.

Unused methods should be removed.

Remove the joinStringArray() method from the Utils class.

This would make it easier for other developers to read.

Wrongian committed:
Merge branch 'branch-Level-7'

Wrongian committed:
Merge branch 'master' into branch-Level-8

Wrongian committed:
Add date and time formatting and parsing

The bot uses the raw dates/times from the user with no regard to
formatting.

Program could need to check certain properties in regards to the
date/time. Date/time display also should be standardized.

Let's,
- Add methods to parse, unparse and display date/times
- Update the test cases
- Change the task classes and chatbot to use the
parsing/unparsing/display methods
- Have the chatbot account for the fact that if the user enters the
date/time format incorrectly
- Account for the fact that the save files can be incorrect

Having the tasks use LocalDateTime objects can make further
implementation such as listing tasks in a specific time period easier.

Wrongian committed:
Merge branch 'branch-Level-8'

Wrongian committed:
Abstract parsing logic into Parser class and make Command classes

Parsing/command logic is in the class of the chatbot.

This makes the responsibility of both parsing and command logic
fall into the chatbot's hands. This also makes parsing and logic highly
coupled with the chatbot.

Let's do,
- Abstract command logic into their corresponding class
- Make Command as a class for shared logic amongst commands
- Abstract parsing logic into its own class
- Fix other minor issues
- Remove unnecessary methods from chatbot class
- Account for command and parsing logic in the chatbot class

The parser should return a Command object and the chatbot should
execute the command. This completely separates their logic.
If there is an error with the parsing, it has to do with the format
of the argument. Whereas if there is a problem with executing the
command, is has not much to do with how the user formatted their
argument. This separates the logic and reduces coupling as well.

Wrongian committed:
Merge branch 'branch-A-MoreOOP'

Wrongian committed:
Divide classes into packages

Classes are in the source folder without being associated with a
package other than the classes in the exceptions folder.

This is poor organisation and quite messy.

Divide classes into folders that make sense:
- Various tasks classes into the tasks folder
- Ui components go into the ui folder
- Utility functions and classes go into utils folder
- Save and storage classes go into the storage folder
- Exceptions go into the exceptions folder
- The various commmand classes go into the Command folder
Also, change cases to accommodate new file paths.

Improves maintainability, readability, scalability as well
as code reusability. This is a relatively smaller project so do not
need that many nested folders.

Wrongian committed:
Merge branch 'branch-A-Packages'

Wrongian committed:
Merge branch 'add-gradle-support' of https://github.com/Wrongian/ip

Wrongian committed:
Edit build.gradle file

Gradle runs seedu.duke.Duke as an application.

This is the wrong class.

Change build.gradle file such that it runs demacia.Demacia instead.

Small change to get gradle to run properly.

Wrongian committed:
Merge branch 'branch-A-Gradle'

Wrongian committed:
Change name of stringIsInt method to stringIsIndex

Method that checks if a string is an index of an array is called
stringIsInt.

This is inaccurate as it an integer can be less than 0 which is not
under the scope of the method.

Change name to stringIsIndex instead to more accurately reflect its
function.

This should help me and developers in the future understand and use
the function.

Wrongian committed:
Change TaskList deleteTask method error checking

During IndexOutOfBoundsException has default error message. This is
displayed to the user.

The user should not have such a technical error message.

Make the error message consistent.

Important that the user gets a user friendly message.

Wrongian committed:
Added JUnit error classes for automated testing

There is no automated testing other than the text-ui testing.

There should be automated testing to streamline the development
process.

Add tests relating to:
- deleteTask and save methods in TaskList class
- toString and save methods in Todo class
- stringIsIndex, formatDateTime and parseDateTime methods in
Utils class
Additionally, add TaskStub class to isolate components for testing.

This allows easy automated testing with JUnit. Tests are mostly written
to cover normal, some boundary and error test cases of these methods.

Wrongian committed:
Merge branch 'branch-A-JUnit'

Wrongian committed:
Modify build.gradle file to build jar files

The plugin shadowJar is not configured correctly to properly and easily
output jar files for release.

Should generate some jar files for releases.

Change configuration for build.gradle.

Have some releases so other users can try it easier.

Wrongian committed:
Merge branch 'branch-A-Jar'

Wrongian committed:
Add checkstyle

Coding standard violations have to be found manually.

This is time consuming.

Add the checkstyle plugin and configuration.

This is to make it easier to check for coding standard violations.
This also allows for automatic checking using gradle.

Wrongian committed:
Fix code to adhere to coding standards

Some code does not adhere to coding standards.

For consistency, readability and other reasons, need to adhere to
coding standards.

Main things to fix:
- Ordering of import statements
- Logical units separated by lines
- Remove unnecessary imports
- White spaces between statements
- Indentation
- Adhere to 120 character line length limit
- Use more enhanced for loops when needed
- Change names of several variables

Focused on making it more readable to a developer looking through
the code for the first time.

Wrongian committed:
Add find command

Users cannot search for tasks.

Users should be able to search for tasks using a command.

Do:
- FindCommand class for the find command
- Change error logic for TaskList class
- Change listTasks method to work for both find and list commands and
update Junit tests
- Add to text ui tests
- Add find command to parser
- Add nameContainsString method to Task class

Use taskList to store the search results as well as the purpose of
the TaskList class in the first place is to store tasks and manipulate
them. The same principle can be applied to the find command as well.

Wrongian committed:
Add Javadocs.

No class have Javadoc comments.

Javadocs make it easier for developers to understand their code,
especially later on.

Add Javadocs to all methods and classes.

Classes and methods follow similar Javadocs style for consistency.

Wrongian committed:
Merge branch 'branch-Level-9'

Wrongian committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/demacia/tasks/TaskList.java

Wrongian committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/demacia/Parser.java
#	src/main/java/demacia/commands/DeadlineCommand.java
#	src/main/java/demacia/commands/EventCommand.java
#	src/main/java/demacia/tasks/Task.java
#	src/main/java/demacia/tasks/TaskList.java
#	src/test/java/demacia/utils/UtilsTest.java

Wrongian committed:
Fix bugs caused by last merge commit and fix coding standards.

Errors due to merge commit exist. Coding standards are not up to par
in other branches that were merged.

Important to fix the bugs and ensure coding standards are acceptable.

Fix residual merge conflict in UtilsTest class. Fix coding standards
such as wrong import order and indentation of case statements. Add
Javadocs to methods that did not have Javadocs.

Fix it in a separate/branch and commit because the last commit was
about merging.

Wrongian committed:
Merge branch 'branch-bugs-and-coding-standards'

Wrongian committed:
Add ability to get total tasks in TaskList

Wrongian committed:
Change text ui messages

Text ui messages displays that there are multiple tasks to the user
when there is only one.

This does mot make sense to the user.

Fix this issue with the various Command classes.

Message the same for consistency.

Wrongian committed:
Amend test cases

Wrongian committed:
Merge branch 'branch-improve-list-message'

TiewJiaLiang commented on src/main/java/demacia/Demacia.java:
```diff
@@ -0,0 +1,73 @@
+import demacia.storage.SaveData;
+import demacia.storage.SaveHandler;
+import demacia.tasks.TaskList;
+import demacia.ui.Terminal;
```
Great Job! You followed the standard by importing classes explicitly

TiewJiaLiang commented on src/main/java/demacia/Demacia.java:
```diff
@@ -0,0 +1,73 @@
+        // get messages from user
+        // todo: enums for the commands
+        // todo: higher order function so that dont have to add horizontal lines to everything
+        boolean isExit = false;
```
Great job in following the standard for boolean variables!

TiewJiaLiang commented on src/main/java/demacia/Parser.java:
```diff
@@ -0,0 +1,172 @@
+     */
+    public static Command parseCommand(String msg) throws IncorrectArgumentFormatException {
+        // hashmap rest of the arguments
+        String[] args = msg.split(" /");
```
Great job in using plural form!

TiewJiaLiang commented on src/main/java/demacia/tasks/TaskList.java:
```diff
@@ -0,0 +1,240 @@
+ * Class representing a collection to store Task objects.
+ */
+public class TaskList implements Saveable {
+    private static final int MAX_TASKS = 100;
```
Great job in following the standards for constant!

TiewJiaLiang reviewed (commented):
Well done! Your codes followed the coding standards :)

Wrongian committed:
Add Javafx

Wrongian committed:
Add styling to DialogBox and MainWindow

Wrongian committed:
Delete old text-ui test files

Wrongian committed:
Add images for chatbot conversation

Wrongian committed:
Amend to fit coding standards

Wrongian committed:
Merge branch 'branch-Level-10'

Siyan-G commented on src/main/java/demacia/storage/SaveHandler.java:
```diff
@@ -0,0 +1,59 @@
+    */
+
+    // hard coded save paths
+    private static final String savePath = "./data/save.txt";
+    private static final String dirPath = "./data";
```
As part of the java coding standard, it is recommended to use screaming snake characters for constants. However, I like the way you declare the constants. Good job!

Siyan-G reviewed (commented):
Overall, a very good jobz with very high code quality and keeping to the java coding standard!!

Siyan-G commented on src/main/java/demacia/tasks/Task.java:
```diff
@@ -0,0 +1,197 @@
+     * @return The Task that from the serialised String.
+     * @throws InvalidSaveException If there are any formatting errors with the save.
+     */
+    public static Task load(String saveString) throws InvalidSaveException {
```
love the way you name the various methods, short, sweet and meaningful

Siyan-G commented on src/main/java/demacia/tasks/Task.java:
```diff
@@ -0,0 +1,197 @@
+     * @param searchString The string to check.
+     * @return Boolean indicating if the name of the Todo contains a String.
+     */
+    public boolean nameContainsString(String searchString) {
```
good use of name to explain the purpose of the method with sufficient details

Siyan-G commented on src/main/java/demacia/exceptions/IncorrectArgumentFormatException.java:
```diff
@@ -0,0 +1,25 @@
+ * Class for the exception relating to having the wrong format for the arguments for
+ * the command such that the chatbot cannot parse it.
+ */
+public class IncorrectArgumentFormatException extends DukeException {
```
good use of names to explain the exception class with sufficient details!

Wrongian committed:
Add assertions to various classes

Wrongian committed:
Merge branch 'branch-A-Assertions'

Wrongian committed:
Change naming for constants

Wrongian committed:
Fix javadocs formatting

Wrongian committed:
Refactor code to remove long methods.

Wrongian committed:
Reduce magic literal usage

Wrongian committed:
Write code to catch SecurityException errors

Wrongian committed:
Merge branch 'branch-A-CodeQuality'

Wrongian committed:
Add gradle.yml

Wrongian committed:
Merge branch 'branch-A-CI'

Wrongian committed:
Merge branch 'master' into branch-A-CodeQuality

Wrongian committed:
Reduce usage of magic literals

Wrongian committed:
Revert "Reduce usage of magic literals"

This reverts commit f9c138d057db825f8234e8540210b8d28955bade.

Wrongian committed:
Reduce magic literals

Wrongian committed:
Merge branch 'branch-A-CodeQuality'

Wrongian committed:
Reimplement find method using stremas

Wrongian committed:
Fix minor error in delete message

Wrongian committed:
Fix test case to have a more accurate delete message

Wrongian committed:
Fix checkstyle violation

Wrongian committed:
Merge pull request #2 from Wrongian/branch-fix_tests

fix: resolve test cases failing

Wrongian committed:
Merge branch 'master' into branch-A-Streams

Wrongian committed:
Merge pull request #3 from Wrongian/branch-A-Streams

feature: add streams in implementations

Wrongian committed:
Add notes as a feature

User has no easy way to input notes for each task.

It would be convenient for the user to input notes to remember
specific things about each task.

Add:
- The ability to save notes to a particular task.
- The ability to get the notes to a particular notes.
- The ability to load notes from the save file.
- The ability to save notes to the save file.

Add new commands and lines to the parser and format that follows
the same format as the rest of the commands to keep it consistent.

Wrongian committed:
Merge branch 'branch-D-Notes'

Wrongian committed:
Add and throw CommandExceptions in various methods

The exceptions of the user entering the wrong format of the command
is not differentiated from the exceptions when the user enters invalid
values.

It makes sense to abstract them as different errors and signal them
as different errors for both for the users and the developers.

Make new kind of Exception called CommandException. This exception
is thrown when a Command execution fails for whatever reason.

This is to differentiate formatting errors of the user from the
wrong command inputs of the user and various other errors. Again, I
do believe this should help the user and developer.

Wrongian committed:
Fix text in dialog box cutting off error

Wrongian committed:
Account for reserved characters when parsing commands

Wrongian committed:
Add functionality for chatbot to tell the user if save/load is wrong

Wrongian committed:
Fix checkstyle errors

Wrongian committed:
Change code to align it more with my chatbot

Wrongian committed:
Merge branch 'branch-clean_various_problems'

Wrongian committed:
Fix event saving bug

Wrongian committed:
Merge branch 'branch-quick_fix'

Wrongian committed:
Add Ui image to docs folder

Wrongian committed:
Merge branch 'branch-UI'

Wrongian committed:
Update github readme

Wrongian committed:
Remove user error message for faulty save

Wrongian committed:
Fix usage instructions for certain commmands

Wrongian committed:
Update readme in /docs

Wrongian committed:
Merge branch 'branch-update_readmes'

Wrongian committed:
Fix ui image not displaying

Wrongian committed:
Merge branch 'branch-update_readme'

Wrongian committed:
Update readme
